File: 3. nacionalna sluzba za zaposljavanje/backend/SluzbaZaposljavanja/src/main/java/com/example/SluzbaZaposljavanja/security/SecurityConfiguration.java
package com.example.SluzbaZaposljavanja.security;

import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.context.annotation.Bean;
import org.springframework.context.annotation.Configuration;
import org.springframework.http.HttpMethod;
import org.springframework.security.authentication.AuthenticationManager;
import org.springframework.security.config.annotation.authentication.builders.AuthenticationManagerBuilder;
import org.springframework.security.config.annotation.method.configuration.EnableGlobalMethodSecurity;
import org.springframework.security.config.annotation.web.builders.HttpSecurity;
import org.springframework.security.config.annotation.web.builders.WebSecurity;
import org.springframework.security.config.annotation.web.configuration.EnableWebSecurity;
import org.springframework.security.config.annotation.web.configuration.WebSecurityConfigurerAdapter;
import org.springframework.security.config.http.SessionCreationPolicy;
import org.springframework.security.core.userdetails.UserDetailsService;
import org.springframework.security.crypto.bcrypt.BCryptPasswordEncoder;
import org.springframework.security.crypto.password.PasswordEncoder;
import org.springframework.security.web.authentication.UsernamePasswordAuthenticationFilter;

@Configuration
@EnableWebSecurity
@EnableGlobalMethodSecurity(prePostEnabled = true)
public class SecurityConfiguration extends WebSecurityConfigurerAdapter{
	
    @Bean
    @Override
    public AuthenticationManager authenticationManagerBean() throws Exception {
        return super.authenticationManagerBean();
    }

    @Bean
    public AuthenticationTokenFilter authenticationTokenFilterBean() throws Exception {
        AuthenticationTokenFilter authenticationTokenFilter = new AuthenticationTokenFilter();
        authenticationTokenFilter.setAuthenticationManager(this.authenticationManagerBean());
        return authenticationTokenFilter;
    }

    @Override
    protected void configure(HttpSecurity httpSecurity) throws Exception{
        httpSecurity.headers().cacheControl().disable();
        httpSecurity.cors();
        httpSecurity.headers().frameOptions().disable();
        httpSecurity.csrf().disable()
                .sessionManagement()
                .sessionCreationPolicy(SessionCreationPolicy.STATELESS)
                .and()
                .authorizeRequests()
                .antMatchers(HttpMethod.POST, "/api/oglasi/**").permitAll()
                .antMatchers(HttpMethod.POST, "/api/oglasi/*").permitAll()
                .antMatchers(HttpMethod.PUT, "/api/oglasi/**").permitAll()
                .antMatchers(HttpMethod.PUT, "/api/oglasi/*").permitAll()
                .antMatchers(HttpMethod.GET, "/api/gradjani/**").permitAll()
                .antMatchers(HttpMethod.GET, "/api/gradjani*").permitAll()
                .antMatchers(HttpMethod.GET, "/api/konkursi/**").permitAll()
                .antMatchers(HttpMethod.GET, "/api/konkursi/*").permitAll()
                .antMatchers(HttpMethod.GET, "/api/oglasi*").permitAll()
                .antMatchers(HttpMethod.GET, "/api/oglasi/**").permitAll()
                .antMatchers(HttpMethod.GET, "/api/firme/**").permitAll()
                .antMatchers(HttpMethod.GET, "/api/poslovi/**").permitAll()
                .antMatchers(HttpMethod.POST, "/api/konkursi").permitAll()
                .antMatchers(HttpMethod.POST, "/api/konkursi/**").permitAll()

                .anyRequest().authenticated();

        httpSecurity.addFilterBefore(authenticationTokenFilterBean(), UsernamePasswordAuthenticationFilter.class);
    }

}

File: spring-auth-integration-demo/src/main/kotlin/io/uvera/springauthintegrationdemo/demo/SecurityConfiguration.kt
package io.uvera.springauthintegrationdemo.demo

import org.springframework.context.annotation.Configuration
import org.springframework.http.HttpMethod
import org.springframework.security.config.annotation.method.configuration.EnableGlobalMethodSecurity
import org.springframework.security.config.annotation.web.builders.HttpSecurity
import org.springframework.security.config.annotation.web.configuration.EnableWebSecurity
import org.springframework.security.config.annotation.web.configuration.WebSecurityConfigurerAdapter
import org.springframework.security.config.http.SessionCreationPolicy
import org.springframework.security.config.web.servlet.CorsDsl
import org.springframework.security.config.web.servlet.invoke
import org.springframework.security.web.authentication.UsernamePasswordAuthenticationFilter
import org.springframework.web.client.RestTemplate
import org.springframework.web.cors.CorsConfiguration
import org.springframework.web.cors.CorsConfigurationSource
import javax.servlet.http.HttpServletRequest

@Configuration
@EnableWebSecurity
@EnableGlobalMethodSecurity(prePostEnabled = true)
class SecurityConfiguration(
    private val jwtAuthEntryPoint: JwtAuthEntryPoint,
    private val restTemplate: RestTemplate,
) : WebSecurityConfigurerAdapter() {

    override fun configure(http: HttpSecurity) {
        http {
            cors {
                configurationSource {
                    corsConfiguration {
                        allowedOrigins = listOf("*")
                        allowedHeaders = listOf("*")
                        allowedMethods = listOf(
                            HttpMethod.GET,
                            HttpMethod.HEAD,
                            HttpMethod.POST,
                            HttpMethod.DELETE,
                            HttpMethod.PUT,
                            HttpMethod.OPTIONS
                        ).map(HttpMethod::toString)
                    }
                }
            }
            csrf { disable() }
            httpBasic { disable() }
            logout { disable() }
            authorizeRequests {
                authorize(anyRequest, authenticated)
            }
            exceptionHandling {
                authenticationEntryPoint = jwtAuthEntryPoint
            }
            sessionManagement {
                sessionCreationPolicy = SessionCreationPolicy.STATELESS
            }
            // idk why but with latest spring some cyclic dependency stuff is going on so we're directly injecting filter
            addFilterBefore<UsernamePasswordAuthenticationFilter>(JwtAuthFilter(restTemplate))
        }
    }
}

inline fun CorsDsl.configurationSource(crossinline block: (HttpServletRequest) -> CorsConfiguration?) {
    this.configurationSource = CorsConfigurationSource {
        block(it)
    }
}

inline fun corsConfiguration(block: CorsConfiguration.() -> Unit) =
    CorsConfiguration().apply {
        block()
    }
